ocaml/typing/typetexp.ml

```diff
@@ -1083,7 +1132,7 @@ and transl_type_var env ~policy ~row_context attrs loc name jkind_annot_opt =
     with Not_found ->
       let jkind =
         try TyVarEnv.lookup_global name |> estimate_type_jkind env
-        with Not_found -> TyVarEnv.new_jkind ~is_named:true policy
+        with Not_found -> TyVarEnv.new_jkind ~is_named:true policy jkind_check
```

Comment:
The logic of `new_jkind` is the right logic for how we should be typing variables. However, rather than only using that logic in the `Not_found` case we need to use it whenever the variable being used doesn't have a kind annotation on its binding site. Otherwise the kind inference will be order-dependent. Consider this example:
```ocaml
type ('f : (value => value) => value) third_order
val foo : 'f third_order -> 'a 'f -> int
```
This signature will type-check because the first occurrence of `'f` is in a checking position. But if we swap the argument order:
```ocaml
type ('f : (value => value) => value) third_order
val foo : 'a 'f -> 'f third_order -> int
```
then it will stop type-checking because the first occurrence of `'f` is in inference position and it will assume `'f` has a second-order kind.

Rather than have the relevant logic in `new_jkind`, we should pull it out into its own function for computing the expected kind of a variable. If the variable is `Not_found` then we should use that kind as the initial kind for the variable. If the variable has an annotation on its binding then we don't need to consider the expected kind. If the variable has no annotation, but has already been seen before then we need to check that the previously created type has the expected kind. This will result in the variable getting the intersection of all its expected kinds -- assuming such an intersection exists.